source/openqasm/parser/antlr/qasm_parser.py

```diff
@@ -0,0 +1,908 @@
+            expression = ctrlModifier.expression()
+        else:
+            modifier = GateModifierName[ctx.getChild(0).getText()]
+            expression = None
+
+        modifier_expression = self.visit(expression) if expression else None
+        return QuantumGateModifier(modifier, modifier_expression)
+
+    def visit(self, tree):
+        try:
+            return super().visit(tree)
+        except:
+            raise
+
+    @span
+    def visitExpressionTerminator(self, ctx: qasm3Parser.ExpressionTerminatorContext):
+        if ctx.Constant():
+            return Constant(ConstantName[ctx.Constant().getText()])
+        elif ctx.Integer():
+            return IntegerLiteral(int(ctx.Integer().getText()))
+        elif ctx.RealNumber():
+            return RealLiteral(float(ctx.RealNumber().getText()))
+        elif ctx.booleanLiteral():
+            return BooleanLiteral(True if ctx.booleanLiteral().getText() == "true" else False)
+        elif ctx.Identifier():
+            return Identifier(ctx.Identifier().getText())
+        elif ctx.StringLiteral():
+            return StringLiteral(ctx.StringLiteral().getText()[1:-1])
+        elif ctx.builtInCall():
+            return self.visit(ctx.builtInCall())
+        elif ctx.externOrSubroutineCall():
+            return self.visit(ctx.externOrSubroutineCall())
+        elif ctx.timingIdentifier():
+            return self.visit(ctx.timingIdentifier())
+        elif ctx.LPAREN():
+            return self.visit(ctx.expression())
+        elif ctx.expressionTerminator():
+            return IndexExpression(
+                self.visit(ctx.expressionTerminator()), self.visit(ctx.expression())
+            )
+
+    @span
+    def visitTimingIdentifier(self, ctx: qasm3Parser.TimingIdentifierContext):
+        if ctx.TimingLiteral():
+            # parse timing literal
+            s = ctx.TimingLiteral().getText()
+            for i, c in enumerate(s):
+                if not c.isdigit() and c != ".":
+                    break
+            return DurationLiteral(float(s[:i]), TimeUnit[s[i:]])
+        elif ctx.Identifier():
+            return DurationOf(
+                target=add_span(Identifier(ctx.Identifier().getText()), get_span(ctx.Identifier()))
+            )
+        else:
+            child_count = ctx.quantumBlock().getChildCount()
+            return DurationOf(
+                target=[
+                    self.visit(ctx.quantumBlock().getChild(i)) for i in range(1, child_count - 1)
+                ]
+            )
+
+    @span
+    def visitUnaryExpression(self, ctx: qasm3Parser.UnaryExpressionContext):
+        return UnaryExpression(
+            UnaryOperator[ctx.unaryOperator().getText()],
+            self.visit(ctx.powerExpression()),
+        )
+
+    @span
+    def visitBuiltInCall(self, ctx: qasm3Parser.BuiltInCallContext):
+        name = ctx.builtInMath().getText() if ctx.builtInMath() else ctx.castOperator().getText()
```

Comment:
I think `castOperator` should not become `FunctionCall`.

Reply:
Agree. Fixed.